python/pyspark/sql/functions.py

```diff
@@ -2200,38 +2200,70 @@ def make_date(year: "ColumnOrName", month: "ColumnOrName", day: "ColumnOrName")
     return Column(jc)
 
 
+@overload
+def date_add(start: "ColumnOrName", days: "ColumnOrName") -> Column:
+    ...
+
+
+@overload
 def date_add(start: "ColumnOrName", days: int) -> Column:
+    ...
+
+
+def date_add(start: "ColumnOrName", days: Union["ColumnOrName", int]) -> Column:
     """
     Returns the date that is `days` days after `start`
 
     .. versionadded:: 1.5.0
 
     Examples
     --------
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['dt'])
+    >>> df = spark.createDataFrame([('2015-04-08', 2,)], ['dt', 'add'])
     >>> df.select(date_add(df.dt, 1).alias('next_date')).collect()
     [Row(next_date=datetime.date(2015, 4, 9))]
+    >>> df.select(date_add(df.dt, df.add.cast(IntegerType())).alias('next_date')).collect()
+    [Row(next_date=datetime.date(2015, 4, 10))]
     """
     sc = SparkContext._active_spark_context
     assert sc is not None and sc._jvm is not None
-    return Column(sc._jvm.functions.date_add(_to_java_column(start), days))
+    return Column(
+        sc._jvm.functions.date_add(
+            _to_java_column(start), days if isinstance(days, int) else _to_java_column(days)
```

Comment:
It might be better to wrap `days` with `lit` outside the block:


```python
days = days if isinstance(days, Column) else lit(days)

return sc._jvm.functions.date_add(_to_java_column(start), _to_java_column(days))
```

Same for the other ones.

Reply:
Sure- I like this better ^^ thanks for the suggestion!

Reply:
I assume that it will still work if I passed a string-name rather than a spark column, right?

Reply:
Alternatively, with `_create_column_from_literal`

```
_create_column_from_literal
days = days._jc if isinstance(days, Column) else _create_column_from_literal(days)

return sc._jvm.functions.date_add(_to_java_column(start), days)
```

Reply:
This also wouldn't allow me to write:

df = df.withColumn("new_date", F.date_add("old_date", "int_column"))

Or am I mistaken?

Reply:
` days = lit(days) if isinstance(days, int) else days

    return Column(
        sc._jvm.functions.date_add(
            _to_java_column(start), _to_java_column(days)
        )
    )`

Reply:
This works ^^ but it's not optimal if you want to avoid the check on the int; is there a benefit you see to running the isinstance on the Column class specifically?

Reply:
You're right. One way or another we have to check on `int`, so maybe the initial approach was OK. Sorry for the confusion.

Reply:
No worries at all! I've amended the code to the above ^^ so keeping a refactor of the 'days' variable- thank you for suggesting this.